webhooks.go

```diff
@@ -0,0 +1,62 @@
+package pagerduty
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"strings"
+)
+
+const (
+	currentSignaturePrefix = "v1="
+)
+
+var (
+	ErrNoValidSignature = errors.New("invalid webhook signature")
+)
+
+// VerifySignature compares the provided signature of a PagerDuty v3 Webhook
+// against the expected value and returns an error if the values do not match.
+//
+// See https://developer.pagerduty.com/docs/webhooks/webhook-signatures/ for more details.
+//
+func VerifySignature(payload []byte, header, secret string) error {
```

Comment:
My concern with this approach is that implementation details of the PagerDuty API leak outside of the package into the consumer's program, specifically needing to know which header to look for and pull it out.

How would you feel about taking the entire `*http.Request` as the first argument, and then the secret as the second? That way the headers and what-not are not exposed to the consumer. After reading the body we could replace it with a `ioutil.NopCloser(bytes.Reader(body))` so that consumers can then read the request body themselves.

Reply:
Thinking more, if we do this we may need another sentinel error for an invalid request (missing / malformed header).

Reply:
Yeah that's not a bad idea.

The intention was definitely to build more helpful public API methods on top of this one.  Was just trying to get some initial code out in public since we have folks asking about how this would be implemented.

Reply:
@theckman : actually I am kind of thinking that the method you are describing might be in addition to this one.  Something like: 

```go
// VerifyWebhookRequest extracts and verifies the webhook signature.
// This method would likely return more detailed errors that differentiate between missing / invalid signatures, etc.
func VerifyWebhookRequest(request *http.Request, secret string) error
```

or
 
```go
// ConstructWebhookEvent verifies and parses a v3 Webhook Event
func ConstructWebhookEvent(request *http.Request, secret string) WebhookEvent, error
```

It still seems useful leave this more raw public method for clients wishing to have more control over this functionality?

Reply:
My personal preference _generally_ is to try and keep the API as small as we can, as it means there's less for us to maintain and honestly less to potentially break by accident in a future update. Text communication is hard, so maybe it'd be best if I try to figure out what assumptions / expectations I think I have:

* I believe the majority of Gophers are working within an environment where the request will be available to them as an `*http.Request`.
* Once it's out of early access, I don't expect the V3 signature format to change much (especially the header) without it incurring a new version designation from PagerDuty.
* Considering it's unlikely to change, someone could lift the verification code out of this project and put it into theirs (or be inspired by it to write their own) if our API signature doesn't work for them.

If we get issues + comments indicating that the `*http.Request` format doesn't work for many, we can look to expose that additional functionality in a future minor release. If we make the API too wide right away, we must release a major version if we realize nobody is using it and no longer wish to support it.